File: 148-sort-list/sort-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* merge(ListNode* low,ListNode* high){
        ListNode *dummy=new ListNode(-1),*temp=dummy;
        while(low!=NULL && high!=NULL){
            if(low->val<=high->val){
                temp->next=low;
                low=low->next;
            }
           else{
                temp->next=high;
                high=high->next;
            }
            temp=temp->next;
        }
        if(low!=NULL){
              temp->next=low;
        }
           else{
             temp->next=high;
        }
        
            return dummy->next;
    }
         ListNode* findMiddle( ListNode* head){
             ListNode* slow=head,*fast=head->next;
            
            while(fast && fast->next){
                slow=slow->next;
                fast=fast->next->next;
            }

            return slow;
         }
     ListNode* mergesort( ListNode* head){
        if(!head || !head->next)return head;
        ListNode* middle = findMiddle(head);
    // ListNode* right = middle->next;
     ListNode* midright=middle->next;
    middle->next = NULL;
   ListNode* left= mergesort(head);
  ListNode*  right = mergesort(midright);

       return  merge(left,right);
     }
    ListNode* sortList(ListNode* head) {
      return  mergesort(head);
    }
};

File: 148-sort-list/sort-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode merge(ListNode head,ListNode mid){
        ListNode ptr=new ListNode(-1);
        ListNode dummy=ptr;
        while(head!=null && mid!=null){
            if(head.val<mid.val){
                ptr.next=head;
                head=head.next;
            }
            else{
              ptr.next=mid;
                mid=mid.next;  
            }
            ptr=ptr.next;
        }
        if(head!=null)ptr.next=head;
        else ptr.next=mid;

        return dummy.next;
    }

    ListNode mergesort(ListNode head){
        if(head==null ||  head.next==null)return head;
        ListNode mid,slow=head,fast=head,prev=null;
        while(fast!=null && fast.next!=null){
             prev=slow;
            slow=slow.next;
            fast=fast.next.next;
        }
        mid=slow;
        prev.next=null;

       head= mergesort(head);
       mid= mergesort(mid);
       return merge(head,mid);

    }
    public ListNode sortList(ListNode head) {
        if(head==null || head.next==null )return head;
        return mergesort(head);
    }
}